### PR TITLE
New Map Mining

### PR DESCRIPTION
**Pull request by LordFowl**
The new map mining system attempts to make mining more of a process than a thing to waste time on. Through the features and functionalities of the mining system miners will be encouraged to progress and in progressing will be forced to actually mine. Because of the way the system is designed mining output in the beginning will be small - much smaller than the output that can be experienced through tactical industrial drilling. However, as the round progresses they will be able of consuming more and more ore, and as a result I have no doubt that they will be able to produce enough ore for the station, God willing.

The bulk of the system focuses around the Mining Vendor, which is soft-ported from /tg/. The vendor does not take cash nor credit, but rather ore-redemption points, which can be accrued through normal processing of ore. (I have not ported the ore redemption machine, but instead have merged its functionality with our already extant smelter). As miners smelt ore, the ore processor racks up points, which they can either leave in the machine or claim by swiping them onto their ID. They can then use their ID to purchase things in the mining vendor. Both machines are compatible with any ID object, not just miner IDs.

Ore redemption proceeds as follows. Each point is per individual ore object processed. Alloying two different ores together should correctly account for their individual worth.

- Sand ore: 1 point
- Coal ore: 2 points
- Iron ore: 4 points
- Phoron ore: 8 points
- Platinum ore: 15 points
- Silver ore: 20 points
- Uranium ore: 25 points
- Gold ore: 30 points
- Metallic hydrogen ore: 30 points
- Diamond ore: 50 points
- Bananium ore: 150 points

Beyond the contents of the mining vendor, which will be detailed below, the system also includes some other miscellaneous changes.

Perhaps most immediate, miners do not start off as well off as they do on the Exodus. Their bigass industrial drills are nothing more than a hole in space (They can now buy them from the mining vendor at a hefty price). They are returned to their Auroran glory, and now spawn with good old pickaxes, like their grand-pappies before them. They also no longer have material or mesons roundstart, but instead an ore radar that functions similarly to the pinpointer, but instead of looking for nuke it looks for nearby ore. They spawn now only with voidsuits - the RIG that once spawned tucked away in their storage is now a mining vendor reward. Finally, there is (tentatively), an additional miner slot, bringing it from 3 miners to 4. It's a big asteroid. The miners also spawn with a single minecart engine and two minecarts, along with some starting track.

Pickaxes and drills are now more distinct than "drills are better than pickaxes, full stop". Some might find even that pickaxes are better for their purpose:

Pickaxes require two hands to be swung effectively - not using two hands triples the amount of time it takes to break through ore. Pickaxes also will deal less damage to mobs when not wielded. Finally, and perhaps most significantly, it takes multiple distinct swings to break through a rock.

Drills require only one hand to be operated. Drills deal less damage than a wielded pickaxe, but a little more than an unwielded pickaxe. Drills require only one click to break through a rock. (There is still a delay before the rock is actually broken).

The main advantage of picks is, that while they may not break through rock as quickly as drills, their more careful excavation may result in a higher net gain of ore. Everytime a pick-wielder successfully completes swing motion, there is a 50% chance that they will produce a rock item, and there is a 25% chance that that rock item is ore of the same type they are mining. Ultimately, it becomes a matter of speed vs profit: Pickaxes require more involvement (Around four clicks per rock), and more time, but produce more resources. Drills require less involvement (Always one click per rock), and typically less time, but also less potential profit.

FINALLY, you can no longer mine multiple turfs at one time with one pick. No longer can you clear out a 3x1 area with ease by somehow striking three rocks at once.

There exist also alternative mining implements, which will be detailed below.

The asteroid itself also looks different:

The asteroid is comprised ultimately of three sections: Near surface, middle, and bottom. Miners begin on the middle section, and can access through the station or their own creativity the bottom section. The near surface section is harder to get to. That said, the near-surface section is abundant with ore, as it has higher spawn chances and near to no actual structures occupying it. The middle and bottom section have roughly the same amount of ore chance, but the bottom section typically has a deal less ore because it is permeated with chasms to space. The asteroid itself is like swiss cheese - covered with both chasms that must be navigated around and caverns that must be traversed.

Possible flora, fauna, and ruins will be added in a subsequent update.

Of minor note, the miners possess a new machine called the RIG press. As its name indicates, its purpose is to convert tools into RIG modules. It only accepts tools that miners either spawn with or can buy from the vendor, and the process is irreversible.

Finally, the big bad motherfucker, Tsar Vendor:

As described above, the vendor operates simply, and can output the following items:
 (Note, this image is a _little_ old, but the actual up to date version regarding prices can be found, surprise surprise, in this very PR.)

Let's just dive in, shall we?

**"Food Ration"**
Pretty self explanatory. It is a pretty robust food item, in that it includes iron, but it's nothing special beyond that.

**"Poster"**
Some bling for the miners, yo.

**"Ore Scanner Pad"**
This is a new item as far as the update goes, but it is still default mining equipment and nothing to scream about. It works similarly to the nuke disc pinpointer, but instead locates the nearest ore turf in a 14 turf radius. Not terrible, but also not terribly useful. Proto-material scanners.

**"5 Red/Green/Yellow Flags"**
In case miners lose their flags, or simply need more. Self explanatory.

**"Ore-bag"**
In case the miners lose their ore bag, or simply need more.

**"Meat Pizza"**
In case miners want to re-enact one of their favourite home videos with a nurse, or just want to celebrate a job well done.

**"Lantern"**
The first of items that miners do not spawn with.  The lantern functions similarly to their headleamp (Which has a big range), but is omnidirectional and warmer. Quickly outphased by the material scanners, but useful for the OCD miner that needs light all around them.

**Shovel**
More replacement equipment for the miners in case they lost their original or just need a new one. The shovel of the future has the added ability of being able to actually dig down, allowing miners to create their own chasms.

**"Stimpack"**
A 20 unit auto-injector of 12 units of hyperzine and 8 units of tramadol.  To be used in case the miner encounters a cave nasty and needs a little edge. THIS ALONG WITH THE SURVIVAL INJECTOR BELOW NEEDS TO BE REVIEWED BY SOMEONE MORE EXPERIENCED WITH MEDICAL. THIS IS PRETTY MUCH A STRAIGHT PORT FROM /TG/ AS IT STANDS NOW.

**"Pickaxe"**
More replacement equipment. Identical to the starting pick.

**"KA White Tracer Rounds"**
A cosmetic upgrade for the kinetic accelerator, this simply adds a white tracer to the otherwise invisible projectile.

**"Compressed rail cartridge"**
Ammunition for the Rapid-Railway-Fabricator, which can also accept RCD ammo if you're a fucking lunatic. This however cannot be used as RCD ammo in reverse.

**"Emergency Floodlight"**
An item that has all the necessary parts to allow miners to construct their own floodlight, identical to the type that xenoarchaeologists spawn with.

**"Premium Cigar"**
A Havanian Cigar freshly rolled, straight from Space Cuba.

**"KA Adjustable Tracer Rounds"**
Identical to the white tracer rounds, except they can be any colour under the sun.

**"Lottery Chip"**
An e-card equipped with its very own scratch-action module, allowing miners to gamble for real money using their fake money.

**"Ripley Paint Kit"**
In the rare case that robotics is useful, mining can purchase this, which spawns a random ripley paint kit for max bling.

**"Material Scanners"**
The first of the real upgrades, this will probably be one of the first purchases of a fresh miner. Otherwise, they function identically to how they do now.

**"Mining Drill"**
A regular run of the mill mining drill. It's differences between itself and the pickaxe are outlined above.

**"Deep Ore Scanner"**
A cheap-ish piece of equipment that is useless without its more expensive brother, the industrial drill. This allows miners to scan for ore that can only be acquired with the industrial drill.

**"Magboots"**
A pair of magboots. They could be useful for scaling chasms.

**"Stimpack Bundle"**
Four stimpacks for the price of 3. A real deal.

**"Autochisel"**
The rare, one of a kind NanoTrasen branded mk2459 AI-assisted autochisel! It allows miners to abandon their dull job of mining and pursue their real dream - sculpting nude statues of lesbay nurses!

**"Jetpack"**
A run of the mill jetpack. It could be useful for traversing and scaling chasms.

**"Drone Drill Upgrade"**
An upgrade for the mining drone's drill, turning it into a sonic jackhammer, which is more effective at both blasting rocks and killing caves lizards.

**"Seismic Charge"**
A costly, single-use device that is nevertheless invaluable for quickly clearing out ore. It will destroy all rocks in a wide radius, leaving the precious ore and any organics in its blast radius mostly unharmed. Can cause moderate damage to the station if used to close, particularly to windows.

**"Industrial Drill Brace"**
An industrial drill brace for the industrial drill. Remember, you need two of these. This item is also remarkable for being one of the few items that, while ordered from the mining vendor, actually arrive on the cargo shuttle.

**"Point Transfer Card"**
500 points for an ID that has 500 mining points on it, that can be swiped on any other ID. For if you're feeling generous to other miners, I guess.

**"Explorer's Webbing"**
An advanced belt that can fit advanced things.

**"Item-Warp Beacon"**
This can be purchased and placed to act as an anchor for the item warp pack.

**"Item-Warp Pack"**
A three-use pack that can be placed on any structure or sufficiently large obj that is unanchored. It will teleport that object with relative precision to the selected warp beacon.

**"Drone Health Upgrade"**
An upgrade that doubles the health of the mining drone.

**"Rapid-Railway-Fabricator"**
This allows for miners to more rapidly construct minetracks. It is fed with compressed rail cartridges.

**"Survival Autoinjector"**
A 35 unit autoinjector ripe with 15 units of tricordrazine, 10 units of inaprovaline, 5 units of oxycodone, and 5 units of methylphenidate. Designed to possibly save a miner from some scary cave shit that would otherwise wreck their day. THIS NEEDS REVIEW FROM SOMEONE WHO KNOWS MORE ABOUT MEDICAL THAN I DO.

**"Brute First-Aid Kit"**
A simple but relatively expensive first aid kit containing brute patches and advanced brute patches, in case a miner needs to do any quick field repair. Does not come with a health scanner.

**"Ore Magnet"**
The ore magnet, when activated, will attract all ore in a 7 turf radius to itself. It can be held either in hand or placed on the ground.

**"Minecart"**
Functioning similarly to the cargo trolley, this is designed to be linked to a minecart engine and does little else except carry ore crates. This arrives on the cargo shuttle.

**"Resonator"**
The resonator functions like a cross between the KA and the seismic charge. It can decimate rocks and creatures after a short delay, but can only support a limited amount of fields at any given time, and requires adjacency to its target.

**"Mining RIG"**
The classic industrial RIG. Identicial to its current version, except it spawns empty.

**"KA Range Increase"**
This upgrade roughly doubles the effective range of the kinetic accelerator.

**"Jaunter"**
A single use item that produces a wormhole that leads whoever enters it into a beacon, much like the hand-held teleporter. Designed for emergency life or death situations.

**"Kinetic Accelerator"**
The Kinetic Accelerator is a ranged piece of work. Decimating rocks instantly with ease from a comfortable distance of 4-5 meters, it also serves as adequate defense in unpressurized environments. It has infinite ammo, but must be periodically recharged with its pump-action generator.

**"1000 credits"**
For the non-betting man. 800 fake money for 1,000 real money.

**"Mass Driver"**
For miners that don't believe in roads (or minecarts). This order spawns an item containing everything a miner needs to create their own mass driver, with internal power source so that it can be used even in the powerless void of space. Designed for flinging crates or miners back to motherbase.

**"Drone Plasma Cutter Upgrade"**
The penultimate upgrade for the mining drones, this gives them a self-recharging plasma cutter.

**"Mining Drone"**
Mining drones are a cross between a miner's best friend and a maintenence drone. They lack the ability to communicate except by printing out status reports of their own writing, and have limited access compared to regular station synthetics. They are however clearly not adjutants of neither station nor AI, but pure miners. Their laws are as follows:

> 	"Serve and obey all NanoTrasen supply department personnel, with priority according to their rank and role."
	"Protect all NanoTrasen supply department personnel to the best of your abilities, with priority according to their rank and role."
	"Preserve your own existence and prevent yourself from coming to harm, so long as doing such does not conflict with any above laws."
	"In absence of any proper instruction, your primary objective is to excavate and collect ore."

Should any mean ghost refuse to become your best friend, any miner can refund their mining drone with no penalty.

**"Minecart Engine"**
Minecarts function almost identically to cargo trolleys, except they are _much_ faster. Their primary limitation however is that they can only traverse turf that has a minetrack on it. Minetracks can be constructed from rods, or from a Rapid Railway Fabricator. The engine pulls the carts, and arrives on the cargo shuttle.

**"Ore Summoner"**
An advanced version of the ore magnet, it simply teleports every ore in range to its location.

**"Plasma Cutter"**
The plasma cutter functions similarly to the kinetic accelerator, except it is not inhibited by pressurized environments, and can cut off limbs, and functions similarly to a welder, and boasts a faster cooldown. It however needs to be recharged like any other energyweapon, and has a limited range. It applies 15 damage of burn to everything it hits, and 15 damage of brute to any living mobs it hits.

**"KA Damage Increase"**
This increases the damage of the kinetic accelerator from 30 to 40 in an unpressurized environment.

**"KA Cooldown Decrease"**
This decreases the cooldown between shots of the KA.

**"Lazarus Injector"**
This fancy autoinjector is designed to be used on simple mobs, and will resurrect them from the brink of death through the power of Christ. If the mob for whatever reason happened to be hostile, it will become perfectly docile.

**"Industrial Drill Head"**
Without this, the industrial drill braces are useless. It functions identically as it does now, but expands to the miners the world of under ground underground mining. It, like its component braces, arrives on the cargo shuttle once ordered from the vendor.

**"Super Resonator"**
The upgraded resonator doubles the amount of fields that can be fielded by the resonance fields, increases their damage, and reduces the amount of time before they detonate.

**"KA AoE Damage"**
This upgrade allows the kinetic accelerator to apply AoE damage to turfs. It does not affect mobs, but merely expands the kinetic accelerator as a mining tool.

**"Diamond Pickaxe"**
It is like regular pickaxe, but better, ya? It's drill counterpart is lost to the annals of RnD.

**"Drone Thermal Drill Upgrade"**
The ultimate upgrade for the mining drone, it permits them to use the auto-recharging thermal drill, thanks to technology sto- redesigned from the Vaurca.

**"Thermal Drill"**
"Pierce the heavens? Son, there won't _be_ any heavens when you're through with it."

All things are subject to change - everything not saved will be lost.

### CODE REVIEW

File: code/game/objects/items/weapons/teleportation.dm [hunk 1]
```diff
@@ -48,7 +48,7 @@ Frequency:
 	if (usr.stat || usr.restrained())
 		return
 	var/turf/current_location = get_turf(usr)//What turf is the user on?
-	if(!current_location||current_location.z==2)//If turf was not found or they're on z level 2.
+	if(!current_location||current_location.z==1)//If turf was not found or they're on z level 2.
```
Comment (Lohikar) [inline code/game/objects/items/weapons/teleportation.dm]: Perhaps this comment should be updated.

File: code/modules/economy/cash.dm [hunk 2]
```diff
@@ -161,4 +161,39 @@ proc/spawn_money(var/sum, spawnloc, mob/living/carbon/human/human_user as mob)
...
+
+/obj/item/weapon/spacecash/ewallet/lotto
+	name = "lottery card"
+	desc = "A scratch-action charge card that contains a variable amount of money."
+	worth = 0
+	var/scratched = 0
+
+/obj/item/weapon/spacecash/ewallet/lotto/attack_self(mob/user)
+	if(!scratched)
+		user << "<span class='notice'>You begin the scratch action process on the charge card. . .</span>"
```
Comment (Lohikar) [inline code/modules/economy/cash.dm]: This sounds a little awkwardly worded, maybe there's a better way to say this.

File: code/modules/mining/machine_rigpress.dm [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	icon = 'icons/obj/stationobjs.dmi'
+	icon_state = "coinpress0"
+	density = 1
+	anchored = 1.0
+	use_power = 1
+	idle_power_usage = 15
+	active_power_usage = 50
+	var/pressing
+
+/obj/machinery/mineral/rigpress/process()
```
Comment (Lohikar) [inline code/modules/mining/machine_rigpress.dm]: Surely you don't need to make this process() just to update the icon.

File: code/modules/mining/machine_vending.dm [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+/obj/machinery/mineral/equipment_vendor
+	name = "mining equipment vendor"
+	desc = "An equipment vendor for miners, points collected at an ore redemption machine can be spent here."
+	icon = 'icons/obj/machines/mining_machines.dmi'
+	icon_state = "mining"
+	density = 1
+	anchored = 1
+	var/obj/item/weapon/card/id/inserted_id
+	var/list/prize_list = list( //keep in order of price
```
Comment (Lohikar) [inline code/modules/mining/machine_vending.dm]: Any reason this can't be global/shared?

File: code/modules/mining/minebot.dm [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+		P.name = "mining report - [customname]"
+		P.info = {"<center><img src = ntlogo.png></center>
+	<center><b><i>NanoTrasen Mining Drone Report</b></i><br>
+	<font size = \"1\">FOR YOUR EYES ONLY</font><hr><br>
+	<font size = \"1\"><font face='Courier New'>[input]</font></font><hr>
+	<font size = \"1\">[status_report]</font><br>
+	<center><img src = barcode[rand(0, 3)].png></center></center>"}
+		P.update_icon()
+		for(var/mob/O in hearers(src, null))
+			O.show_message("\icon[src] <span class = 'notice'>The [usr] pings, \"[P.name] ready for review\", and happily disgorges a small printout.</span>", 2)
```
Comment (Lohikar) [inline code/modules/mining/minebot.dm]: Visible message?

File: code/modules/multiz/movement.dm [hunk 6]
```diff
@@ -119,7 +119,7 @@
 		return
 
 	if(can_fall())
-		handle_fall(below)
+		handle_fall(below,)
```
Comment (Lohikar) [inline code/modules/multiz/movement.dm]: Rogue comma?

File: code/modules/projectiles/guns/energy/special.dm [hunk 7]
```diff
@@ -424,7 +426,49 @@
 					"<span class='danger'>You hear a low pulsing roar!</span>"
 					)
 	is_charging = 1
-	sleep(60)
+	sleep(40)
+	is_charging = 0
+	if(!istype(user.get_active_hand(), src))
+		return
+	msg_admin_attack("[key_name_admin(user)] shot with \a [src.type] [key_name_admin(src)]'s target (<A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[src.x];Y=[src.y];Z=[src.z]'>JMP</a>)")
```
Comment (Lohikar) [inline code/modules/projectiles/guns/energy/special.dm]: Pretty sure you need to write to the mobs' attack log vars too.

File: code/modules/mining/machine_vending.dm (hunk 4 above)
Comment (LordFowl) [inline code/modules/mining/machine_vending.dm]: I planned at one point to have research be able to upgrade the machines for even higher tiers items, but I never created enough content. Could be global at this point, yea.

File: code/modules/mining/machine_rigpress.dm (hunk 3 above)
Comment (LordFowl) [inline code/modules/mining/machine_rigpress.dm]: It processes already as a machine, does it not? That is why it inherits from the base machine process.

Comment (Lohikar) [inline code/modules/mining/machine_rigpress.dm]: The machinery process will automatically stop a machine from processing if it does not override `process()`. Power is calculated separately.

File: code/modules/projectiles/guns/energy/special.dm (hunk 7 above)
Comment (LordFowl) [inline code/modules/projectiles/guns/energy/special.dm]: A flaw with no_attack_log in general, probably better suited for a seperate bugfix PR.

File: code/modules/mining/machine_vending.dm [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+/datum/data/mining_equipment/New(name, path, cost, shuttle)
+	src.equipment_name = name
+	src.equipment_path = path
+	src.cost = cost
+	src.shuttle = shuttle
+
+/obj/item/weapon/circuitboard/machine/mining_equipment_vendor
+	name = "circuit board (Mining Equipment Vendor)"
+	build_path = /obj/machinery/mineral/equipment_vendor
+	origin_tech = "programming=1;engineering=3"
```
Comment (Lohikar) [inline code/modules/mining/machine_vending.dm]: This'll need converting.

File: code/modules/mining/mine_items.dm [hunk 9]
```diff
@@ -35,32 +34,146 @@
 	icon_state = "lantern"
 	desc = "A mining lantern."
 	brightness_on = 4			// luminosity when on
-	light_power = 0.75
+	light_power = 1
+	light_range = 6
```
Comment (Lohikar) [inline code/modules/mining/mine_items.dm]: `light_range` will be overridden by `brightness_on` when the lantern is turned on.

File: code/modules/mining/mine_items.dm [hunk 10]
```diff
@@ -249,3 +387,924 @@
...
+	icon_state = "track15"
+	density = 0
+	anchored = 1.0
+	w_class = 3
+	layer = 2.44
+
+/obj/structure/track/initialize()
+	..()
+	for(var/obj/structure/track/T in src.loc)
+		if(T != src)
```
Comment (Lohikar) [inline code/modules/mining/mine_items.dm]: Perhaps this loop could instead do:
 ```
if (locate(/obj/structure/track) in loc) 
    qdel(src)
```

File: code/modules/mining/mine_items.dm [hunk 11]
```diff
@@ -249,3 +387,924 @@
...
+	switch(severity)
+		if(1.0)
+			qdel(src)
+			return
+		if(2.0)
+			qdel(src)
+			return
+		if(3.0)
+			return
+		else
```
Comment (Lohikar) [inline code/modules/mining/mine_items.dm]: Trailing else.

File: code/modules/mining/mine_items.dm [hunk 12]
```diff
@@ -249,3 +387,924 @@
...
+		var/obj/item/weapon/weldingtool/WT = C
+		if(WT.remove_fuel(0, user))
+			user << "<span class='notice'>Slicing apart connectors ...</span>"
+		getFromPool(/obj/item/stack/rods, src.loc)
+		qdel(src)
+
+	return
+
+/obj/structure/track/proc/updateOverlays()
+	spawn(1)
```
Comment (Lohikar) [inline code/modules/mining/mine_items.dm]: Why spawn(1)? Might as well just put `set waitfor = FALSE` at the top, unless you *need* the 1ds delay.

File: code/modules/mining/mine_items.dm [hunk 13]
```diff
@@ -249,3 +387,924 @@
...
+	load_offset_x = 0
+	mob_offset_y = 15
+	active_engines = 1
+
+	light_power = 1
+	light_range = 6
+	light_wedge = LIGHT_WIDE
+	light_color = LIGHT_COLOR_FIRE
+
+/obj/vehicle/train/cargo/engine/mining/New()
```
Comment (Lohikar) [inline code/modules/mining/mine_items.dm]: Is this mapped in? If so, maybe this should be `initialize()`.

File: code/modules/mining/mine_items.dm [hunk 14]
```diff
@@ -249,3 +387,924 @@
...
+	name = "jaunt tunnel"
+	icon = 'icons/obj/objects.dmi'
+	icon_state = "bhole3"
+	desc = "A stable hole in the universe made by a wormhole jaunter. Turbulent doesn't even begin to describe how rough passage through one of these is, but at least it will always get you somewhere near a beacon."
+
+/obj/effect/portal/wormhole/jaunt_tunnel/teleport(atom/movable/M)
+	if(M.anchored || istype(M, /obj/effect))
+		return
+	single_spark(M.loc)
+	if(istype(M, /atom/movable))
```
Comment (Lohikar) [inline code/modules/mining/mine_items.dm]: Could just be `istype(M)`.

File: code/modules/mining/mine_turfs.dm [hunk 15]
```diff
@@ -531,14 +648,35 @@
...
+		/obj/random/action_figure = 0.1)
+		var/ore_path = pickweight(ore)
+		if(ore)
+			new ore_path(src)
+			user << "<span class='notice'>You unearth something amidst the sand!</span>"
+
+	if(dug <= 10)
+		dug += 1
+	else
+		if(GetBelow(src))
```
Comment (Lohikar) [inline code/modules/mining/mine_turfs.dm]: This is handled by ChangeTurf automatically, no?

**Comment by Fire-and-Glory** [comment]
>"Lazarus Injector"
This fancy autoinjector is designed to be used on simple mobs, and will resurrect them from the brink of death through the power of Christ. If the mob for whatever reason happened to be hostile, it will become perfectly docile.

>perfectly docile

i've been lied to

**Pull request merged**